tikv-client/src/main/java/com/pingcap/tikv/PDClient.java

```diff
@@ -95,34 +97,93 @@ public class PDClient extends AbstractGRPCClient<PDBlockingStub, PDStub>
+  private PDClient(TiConfiguration conf, RequestKeyCodec keyCodec, ChannelFactory channelFactory) {
     super(conf, channelFactory);
     initCluster();
     this.blockingStub = getBlockingStub();
     this.asyncStub = getAsyncStub();
+    this.upstreamPDClient = createUpstreamPDClient(keyCodec);
+  }
+
+  private org.tikv.common.ReadOnlyPDClient createUpstreamPDClient(RequestKeyCodec keyCodec) {
+    org.tikv.common.TiConfiguration tikvConf = convertTiConfiguration(conf);
+    org.tikv.common.util.ChannelFactory channelFactory = convertChannelFactory(conf, tikvConf);
+    return org.tikv.common.PDClient.create(tikvConf, keyCodec, channelFactory);
+  }
+
+  private org.tikv.common.TiConfiguration convertTiConfiguration(
+      com.pingcap.tikv.TiConfiguration conf) {
+    org.tikv.common.TiConfiguration tikvConf =
+        org.tikv.common.TiConfiguration.createDefault(
+            conf.getPdAddrs().stream().map(Objects::toString).collect(Collectors.joining(",")));
+    // s --> ms
+    tikvConf.setTimeout(conf.getTimeout() * 1000L);
+    tikvConf.setKvClientConcurrency(conf.getKvClientConcurrency());
+    tikvConf.setIsolationLevel(conf.getIsolationLevel());
+    tikvConf.setCommandPriority(conf.getCommandPriority());
+    return tikvConf;
+  }
+
+  private org.tikv.common.util.ChannelFactory convertChannelFactory(
+      com.pingcap.tikv.TiConfiguration conf, org.tikv.common.TiConfiguration tikvConf) {
+    org.tikv.common.util.ChannelFactory tikvFactory = null;
+    if (conf.isTlsEnable()) {
+      if (conf.isJksEnable()) {
+        tikvFactory =
+            new org.tikv.common.util.ChannelFactory(
+                conf.getMaxFrameSize(),
+                tikvConf.getKeepaliveTime(),
+                tikvConf.getKeepaliveTimeout(),
+                tikvConf.getIdleTimeout(),
+                conf.getConnRecycleTime(),
+                conf.getCertReloadInterval(),
+                conf.getJksKeyPath(),
+                conf.getJksKeyPassword(),
+                conf.getJksTrustPath(),
+                conf.getJksTrustPassword());
+      } else {
+        tikvFactory =
+            new org.tikv.common.util.ChannelFactory(
+                conf.getMaxFrameSize(),
+                tikvConf.getKeepaliveTime(),
+                tikvConf.getKeepaliveTimeout(),
+                tikvConf.getIdleTimeout(),
+                conf.getConnRecycleTime(),
+                conf.getCertReloadInterval(),
+                conf.getTrustCertCollectionFile(),
+                conf.getKeyCertChainFile(),
+                conf.getKeyFile());
+      }
+    } else {
+      tikvFactory =
+          new org.tikv.common.util.ChannelFactory(
+              conf.getMaxFrameSize(),
+              tikvConf.getKeepaliveTime(),
+              tikvConf.getKeepaliveTimeout(),
+              tikvConf.getIdleTimeout());
+    }
+    return tikvFactory;
+  }
+
+  public Long getClusterId() {
+    return header.getClusterId();
```

Comment:
Why we getClusterId from header

Reply:
Implements the ReadOnlyPDClient interface, which has the same behavior as upstream